### DISCUSSION TITLE
Road To 1.0

### DISCUSSION THREAD

**Discussion by jkeiser** [simdjson/simdjson#1405] [General]
We've finished a lot of our goals up to this point. 1.0 seems like it could happen soon. Let's list what 1.0 means now in our minds.

**Comment by jkeiser** [comment]
I think simdjson 1.0 should be:

- **The fastest** C++ JSON reading library.
- **Bug free** and well-tested.
- **Intuitive** to use, where the easy thing is the right thing.
- **Safe** so incorrect API use surfaces early in compile or first run, and not as production bugs.
- **Easily integrated** into existing C++ projects.
- **Fully featured** and able to replace other JSON libraries in most situations.

A note about the last criteria: both of us would be fine if other JSON libraries learned from simdjson and got fast, and simdjson wasn't needed anymore. But the best way to make sure simdjson is heard 'round the world is for more people to use it and see the proof that easy and fast *is* possible.

###  What's Left

Overall, we're very close to that criteria. What I think is left:

* On Demand MVP -- There are still a couple of safety issues around arrays and objects I'd like to clear up. 
* JSON builder -- without this we're not fully featured enough to replace other JSON libraries. I think we should do this via the DOM interface. I don't think we need to support edit (read-then-write) capability.

Stretch:
* DOM Deemphasis -- I'd like to add recursive tests (like building a DOM) for On Demand. If it is already reasonably performant, one or two simple methods should be enough to make it easy. In this model, we would replace most of the documentation with "use On Demand," and note that if you need minification or JSON Path, DOM is still the way.
* DOM Deprecation -- When On Demand supports minification and JSON Path, DOM can be *deprecated.* These won't be hard to build and aren't particularly tricky (no weird design interactions)--but like all other things, they take time.

**Reply by jkeiser** [reply]
What really drives the above principles:

> Top performance and slick usability make simdjson easy to pick up.
> 
> Rock-solid reliability and features mean you never have to put it down.

**Reply by jkeiser** [reply]
The flip side: you can make a case that 1.0 should just be the best c++ read api, and release a write api when we have one nicely baked (and 2.0 it, perhaps). It'll limit the reach somewhat perhaps, but it's not like you can't release 2.0 when the write api is done!

If we go that direction, we pretty much just have extra on demand array/object safety, and analysis of recursion support / deemphasizing dom to pare down our api surface and choices users have to make. I can't think of any other *features* we'd want.

**Comment by lemire** [comment]
If we deprecate/de-emphasized the DOM API then this makes some work obsolete:
https://github.com/simdjson/simdjson/issues/1413

I think we need to take a decision. It is going to impact many things if we retire the DOM API... it would trim a lot of work. We would still keep DOM, I think, for the streaming approach (ndjson).


Honestly, at this point, I am thinking that it sounds good. 

We don't want to leave people who rely on the DOM API in the cold... and given that we still want to have it for ndjson (at least for the time being)... so we would probably keep it but more as 'legacy'. E.g., all our documentation (for 1.0) could be around On Demand (again, except for ndjson).

The understanding is that all ongoing development would be on On Demand and the DOM would be just on support mode (bug fixes).

**Comment by jkeiser** [comment]
Yep. ndjson has obvious synergy with on demand, too, and could probably be integrated without much trouble.

**Comment by jkeiser** [comment]
From discussions, it sounds like we're converging on releasing without json builder, deemphasizing the DOM API in favor of On Demand, and not necessarily adding json path or printing to On Demand. Trying to think of criteria for a release, here are some questions worth reflecting on when we decide which features we should include and whether to release:

Will the features be:

* Exciting and unique enough to interest people?
* Useful and complete enough to use all the way to production?
* Fully polished and well-engineered, with no "broken windows" we'd be ashamed of and might turn users off?
* Rock-solid reliable, bug free and fully tested?

I think the answer to all these is yes.

* Ridiculous performance wrapped in an intuitive API is very unique and exciting. 
* A fast read API that can traverse and read any JSON document is enough to bring a huge category of people all the way through production (even though a JSON builder and On Demand JSON path support will attract even more).
* I'm proud of how stage 1, stage 2 and the APIs all turned out. With the final tweaks to On Demand, I'll feel confident it's reliable, well-rounded and safe to use without fear.
* It's damn well-tested. More fuzzing for On Demand wouldn't be amiss, but the test suites are super comprehensive.

**Reply by lemire** [reply]
To be clearer, I think you are saying that if we releasing without json builder, deemphasizes the DOM API in favor of On Demand, and not add json path or printing to On Demand, we will meeting your criteria.

If so I agree.

**Comment by vadi2** [comment]
The library I am most certain is excellently tested, very well written, and very fast.

My 2c on the intuitive and easily integrated - as an intermediate C++ dev who stays out of templates and the dark arts that lie beyond for good reasons[1], the library right now is focused towards advanced C++ developers rather than the general skillset.

My journey goes like this: the CMake integration, albeit somewhat non-standard, worked great.

Having read `basics.md`, the first / obvious question that basics didn't answer was what kind of pattern does the library use when testing an object for the presence of a key - do you test first and try to read, or do you read first and check the error? The answer is now answered in https://github.com/simdjson/simdjson/issues/1424.

Then there is the new ondemand API - given how the docs touted it being amazing for reading selective values, I assumed since I needed to read the JSON entirely it was not for me so went with the old way before finally asking the question https://github.com/simdjson/simdjson/issues/1436 (answer is - use ondemand still in this case). The docs mislead me here a bit, so once I get the old way working, I'll rewrite it in the new way (if it supports an exception-free workflow).

So, armed with the knowledge from `basics.md` I rewrote the existing Qt-based parsing to be simdjson-based. Was happily working through compile-based errors, thought I did the job, compiled, and got a ton of link errors - with no references to code at all. That was pretty nasty https://github.com/simdjson/simdjson/issues/1435. Turns out I was passing QStrings, ints, floats to simd and that's not good.

My entire codebase does not work with the exception based pattern - and I appreciate the fact that simdjson doesn't force them either - but finding a way to work without them was challenging. At the end of the day the undocumented `SIMDJSON_EXCEPTIONS` was very enlightening and would have saved me a heap of work from the start. After some effort, I did get simdjson compiling and working 👍.

[1] - it's an open-source, hobby project that welcomes anyone who wants to come code, this means devs from other languages. Setting the cap at templates & beyond saves a lot of C++ spec reading and makes the project reasonably accessible, while reaping the benefits of a fast language. simdjson's doxygen documentation for this reason is also tough to read as it assumes templates+ level of knowledge.

**Reply by lemire** [reply]
Because we are pivoting to On Demand as our primary front-end, we are going to have to rewrite basics.md. Would you be willing to serve as a reviewer, to provide us feedback on the new documentation?

**Reply by vadi2** [reply]
Happy to help

**Reply by lemire** [reply]
@vadi2 Or `basics.md` as it stands in our main branch is what is planned for release 1.0. Would you review it ?

**Reply by vadi2** [reply]
I've reviewed it and it looks great, thank you!

**Reply by lemire** [reply]
Thanks for the feedback.

**Comment by lemire** [comment]
[I have laid out a project for 1.0](https://github.com/simdjson/simdjson/projects/6).

**Comment by lemire** [comment]
For a status report.

Version 1.0 is effectively feature complete as far as I am concerned.

One thing that I removed was the 'no padding' access. @jkeiser had done a ton of work on the issue, which I picked up and tuned, but I don't want to do a major release with it. I kept finding new bugs and the performance cost could be high.

The current plan is to work for a few weeks with @NicolasJiaxin and others, making sure everything is solid... especially with respect to the new On Demand front-end... and then release 1.0 at the end of the summer, hopefully as a really super solid library that can go into production with confidence.
